external/fv3fit/fv3fit/emulation/thermobasis/emulator.py

```diff
@@ -0,0 +1,375 @@
+    as e.g. :py:class:`V1QCModel`.
+    """
+
+    def __init__(
+        self, config: Config,
+    ):
+        self.config = config
+        self.model = get_model(config)
+        self.optimizer = tf.optimizers.SGD(
+            learning_rate=config.learning_rate, momentum=config.momentum
+        )
+        self.output_variables: Sequence[str] = get_prognostic_variables()
+        self.extra_inputs = config.extra_input_variables
+        self._step = 0
+        self.logger = LoggerList([TBLogger(), ConsoleLogger(), WandBLogger()])
+
+        self.logger = LoggerList([TBLogger(), ConsoleLogger()])
+
+        if config.wandb_logger:
+            self.logger.loggers.append(WandBLogger())
+
+    @property
+    def input_variables(self):
+        return self.config.input_variables
+
+    def batch_to_specific_humidity_basis(self, x):
+        return batch_to_specific_humidity_basis(x, self.extra_inputs)
+
+    def step(self, in_, out):
+
+        with tf.GradientTape() as tape:
+            loss, info = self.get_loss(in_, out)
+
+        vars = self.model.trainable_variables
+        grads = tape.gradient(loss, vars)
+
+        if numpy.isnan(loss.numpy()):
+            raise ValueError("Loss is NaN")
+        self.optimizer.apply_gradients(zip(grads, vars))
+
+        for key in info:
+            tf.summary.scalar(key, info[key], step=self._step)
+        self._step += 1
+        return info
+
+    def get_loss(self, in_, out):
+        return self.config.target.loss(self.model(in_), out)
+
+    def score(self, d: tf.data.Dataset):
+        losses = defaultdict(list)
+        for x, y in d.batch(10_000):
+            in_ = self.batch_to_specific_humidity_basis(x)
+            out = batch_to_specific_humidity_basis(y)
+            _, info = self.get_loss(in_, out)
+            for key in info:
+                losses[key].append(info[key])
+        loss_epoch_test = average(losses)
+        return loss_epoch_test
+
+    def batch_fit(self, d: tf.data.Dataset, validation_data=None):
+        """
+
+        Args:
+            d: a unbatched dataset of tensors. batching is controlled by the
+                routines in side this function
+            validation_data: an unbatched validation dataset
+        """
+
+        if not self.model.scalers_fitted:
+            argsin, argsout = next(iter(d.batch(10_000)))
+            # calls .build on any layers
+            # self.model(argsin)
+            in_ = self.batch_to_specific_humidity_basis(argsin)
+            out = batch_to_specific_humidity_basis(argsout)
+            self.model.fit_scalers(in_, out)
+
+        for i in range(self.config.epochs):
+            logging.info(f"Epoch {i+1}")
+            train_loss = defaultdict(lambda: [])
+            for x, y in d.batch(self.config.batch_size):
+                in_ = self.batch_to_specific_humidity_basis(x)
+                out = batch_to_specific_humidity_basis(y)
```

Comment:
Feels a bit cumbersome to have all the tf.ds batch -> basis handling within the model.  Limitations of tf dataset usage.  Maybe some sort of `BasisIterator` could be useful down the line?

Reply:
I wouldn't say `Trainer` is the "model" it is the training loop. I see it is an intermediary between the dataset and the actual ML models, so this data transformations seems appropriate. I started this code by pushing this translation logic mostly into the data pipeline, but found this tightly coupled the code in `fv3fit.emulation.data` to the particular ML problem. E.g adding a new input variable required changes to both `fv3fit.emulation.data` etc.